### DISCUSSION TITLE
Proxy videos through Tor, 2 questions

### DISCUSSION THREAD

**Discussion by DeltaEpsilon19498** [FreeTubeApp/FreeTube#4563] [General]
Thanks to https://github.com/FreeTubeApp/FreeTube/discussions/4125, I now know how to proxy videos through Tor. That is great! However, I have two questions.

1. I also use the invidious API. Invidious has onion instances (https://docs.invidious.io/instances/#tor-onion-services). For freetube, the current invidious instance will be randomized upon startup, but it seems to randomize between clearnet instances. How do I change the settings so it only chooses between the onion instances? Tor can work via the clearnet instances, but I would prefer to not have to trust the exit nodes.

2. Since tor is very slow for videos, if I open up a video, I like to pause it in my window and do something else in order to let the video load up. However, if I do this in Freetube, it seems to only store a minute ahead in the cache until it stops. I do not want to pause and unpause the video every minute. How do I change it so that it will load the entire video in the cache before I start playing it?

If it matters, I am using FreeTube on Linux Mint, not Windows.

Thank you for your help.

**Comment by absidue** [comment]
> How do I change the settings so it only chooses between the onion instances?

There is no setting, FreeTube intentionally filters out onion instances, as they are useless to 99.9% of FreeTube users. If you want an onion instance (most of them are just onion addresses that point to the same servers as the clearnet URLs), you can copy one from https://api.invidious.io and enter them into the instance field and then set it to default, so it always uses that URL.

> How do I change it so that it will load the entire video in the cache before I start playing it?

No setting for that, there is a feature request open here https://github.com/FreeTubeApp/FreeTube/issues/678 for it, which hasn't been implemented because the player that FreeTube uses, doesn't let you customise the buffer size.

**Reply by absidue** [reply]
As stuff is going to be unbearably slow through tor, you are better off either downloading the video with one of the dedicated downloading programs (similar to buffering the whole video, except that you won't be using up all of your computers RAM) or setting the quality low enough, so that you are at least giving it a chance at being able to load stuff.

**Reply by DeltaEpsilon19498** [reply]
Thank you. Is there any configuration file that I could add the onion instances to, as opposed to choosing 1 onion instance?

Can the second problem be bypassed by using an external player like vlc or mpv somehow?

**Reply by DeltaEpsilon19498** [reply]
Also, since yt-dlp apparently does not support invidious, do you have any suggestions on dedicated downloading programs?

**Reply by absidue** [reply]
> Thank you. Is there any configuration file that I could add the onion instances to, as opposed to choosing 1 onion instance?

No, FreeTube doesn't have any hidden settings, the settings database expects stuff in the same format as the UI and the rest of the code, so in this case if you tried adding multiple default instances, you would just break FreeTube.

For your niche use case your best best is to clone the code and create a custom build for yourself that reverses the onion check (so only including onion, instead of filtering them out) when FreeTube fetches the instances list from https://api.invidious.io.

> Can the second problem be bypassed by using an external player like vlc or mpv somehow?

You can use external players if you want, but they either use yt-dlp or talk to YouTube directly themselves in the background so will download directly from YouTube, just like yt-dlp and youtube-dl.

> Also, since yt-dlp apparently does not support invidious, do you have any suggestions on dedicated downloading programs?

Just use yt-dlp with tor or a VPN, I know that isn't a satisfactory answer for you, but your use case and threat model is a lot more extreme than most people's and most programs are designed to cater for the masses. So to satisfy your use case and threat model, you are going to have to get your hands dirty and edit code.

**Reply by DeltaEpsilon19498** [reply]
Thank you. If I use yt-dlp through tor instead of the tor browser itself, what are the risks of leaking a unique browser fingerprint?

Also, suppose Google were to try to embed some sort of malware/spyware into the yt-dlp downloads of anyone who downloads via tor. How could I defend myself against such a possibility?

Or suppose Google is instructed by a three-letter agency to embed spyware into the downloads via tor of a certain video that the three-letter agencies might not like, such as Professor Mearsheimer's lecture video on "why ukraine is the west's fault." How can I protect myself from similar things, if I consume a video that one or multiple world powers do not like, on a topic that is sensitive to one or multiple of them?

Also, how do I make sure that my yt-dlp download from youtube / invidious is not somehow injected with spyware through a malicious tor node?

Thank you.
